haskell-src/Concordium/Utils/BinarySearch.hs

```diff
@@ -0,0 +1,63 @@
+-- |This module implements binary search on vectors.
+module Concordium.Utils.BinarySearch where
+
+import qualified Data.Vector as Vec
+
+-- |Perform a binary search on a sorted vector, returning the index and the value, if found.
+-- The first argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearchI :: (Ord k) => (a -> k) -> Vec.Vector a -> k -> Maybe (Int, a)
+binarySearchI valToKey vec key = search 0 (Vec.length vec - 1)
```

Comment:
It's very sad that this is necessary. There is an implementation of binary search https://hackage.haskell.org/package/vector-algorithms-0.8.0.4/docs/Data-Vector-Algorithms-Search.html Why was that not used?

If this implementation is used in performance sensitive contexts it should be benchmarked.